les/odr_requests.go

```diff
@@ -445,32 +454,36 @@ func (r *ChtRequest) Validate(db ethdb.Database, msg *Msg) error {
 		}
 
 		// Verify the CHT
-		var encNumber [8]byte
-		binary.BigEndian.PutUint64(encNumber[:], r.BlockNum)
-
-		reads := &readTraceDB{db: nodeSet}
-		value, _, err := trie.VerifyProof(r.ChtRoot, encNumber[:], reads)
-		if err != nil {
-			return fmt.Errorf("merkle proof verification failed: %v", err)
-		}
-		if len(reads.reads) != nodeSet.KeyCount() {
-			return errUselessNodes
-		}
-
+		// Note: For untrusted CHT request, there is no proof response but
+		// header data.
 		var node light.ChtNode
-		if err := rlp.DecodeBytes(value, &node); err != nil {
-			return err
-		}
-		if node.Hash != header.Hash() {
-			return errCHTHashMismatch
-		}
-		if r.BlockNum != header.Number.Uint64() {
-			return errCHTNumberMismatch
+		if !r.Untrusted {
+			var encNumber [8]byte
+			binary.BigEndian.PutUint64(encNumber[:], r.BlockNum)
+
+			reads := &readTraceDB{db: nodeSet}
+			value, _, err := trie.VerifyProof(r.ChtRoot, encNumber[:], reads)
+			if err != nil {
+				return fmt.Errorf("merkle proof verification failed: %v", err)
+			}
+			if len(reads.reads) != nodeSet.KeyCount() {
+				return errUselessNodes
+			}
+
+			if err := rlp.DecodeBytes(value, &node); err != nil {
+				return err
+			}
+			if node.Hash != header.Hash() {
+				return errCHTHashMismatch
+			}
+			if r.BlockNum != header.Number.Uint64() {
+				return errCHTNumberMismatch
+			}
 		}
 		// Verifications passed, store and return
 		r.Header = header
 		r.Proof = nodeSet
-		r.Td = node.Td
+		r.Td = node.Td // For untrusted request, td here is nil, todo improve the les/2 protocol
```

Comment:
Yes I agree even for the `untrusted` style header retrieval, we can also verify the response with a `self given cht root hash`. But the problem here is: in the Les2, we can't get the proof data in the response(only has the corresponding header data), since we don't have a `chtRoot` for to include an unstable header. That's the reason I said "we have to upgrade the protocol to fix it"

Reply:
And for the les1, we even can't get the header data in the response because of missing the corresponding `chtRoot`. So it means we can't support this type of syncing for `les1`

Reply:
ok, we can figure this out later, it is good for now